recipe/meta.yaml

```diff
@@ -1,119 +1,138 @@
 {% set version = "0.16.0" %}
 {% set filename = "apache-arrow-" + version + ".tar.gz" %}
 {% set checksum = "261992de4029a1593195ff4000501503bd403146471b3168bd2cc414ad0fb7f5" %}
+{% set number = "2" %}
+{% set cuda_enabled = cuda_compiler_version is not undefined and cuda_compiler_version == '9.2' %}
+{% set cuda_ext = "-cuda" if cuda_enabled else "" %}
 
 package:
-  name: arrow-cpp
-  version: {{ version }}
+  name: arrow-cpp-ext
 
 source:
   fn: {{ filename }}
   url: https://dist.apache.org/repos/dist/release/arrow/arrow-{{ version }}/{{ filename }}
   sha256: {{ checksum }}
 
 build:
-  number: 1
-  skip: true  # [win and vc<14]
+  skip: true  # [(win and vc<14) or (linux and cuda_compiler_version is not undefined and cuda_compiler_version != '9.2' and cuda_compiler_version != 'None')]
+  number: {{ number }}
   run_exports:
-    - {{ pin_subpackage("arrow-cpp", max_pin="x.x.x") }}
+    - {{ pin_subpackage("arrow-cpp" + cuda_ext, max_pin="x.x.x") }}
 
-requirements:
-  build:
-    - cmake
-    - autoconf  # [unix]
-    - ninja
-    - {{ compiler('c') }}
-    - {{ compiler('cxx') }}
-  host:
-    - aws-sdk-cpp
-    - boost-cpp
-    - brotli
-    - bzip2
-    - c-ares
-    - gflags
-    - glog
-    - grpc-cpp
-    - libprotobuf
-    - clangdev 7.*
-    - llvmdev 7.*
-    - lz4-c
-    - numpy 1.16.*
-    - orc  # [unix]
-    - python
-    - rapidjson
-    - re2
-    - snappy
-    - thrift-cpp >=0.11
-    - zlib
-    - zstd
+outputs:
+  - name: arrow-cpp{{ cuda_ext }}
```

Comment:
Historically we have made a separate mutex metapackage for this kind of selection. That way the user can choose what this should be pretty easily. Also it is less dependent on the solver this way.

Reply:
This mutex metapackage can also be useful for [aligning other downstream packages]( https://github.com/conda-forge/ucx-split-feedstock/blob/84dbdb4d4d58b7cbbe23f15803490658f95f7470/recipe/meta.yaml#L54-L55 ).

Reply:
I think https://github.com/conda-forge/arrow-cpp-feedstock/pull/123#issuecomment-601320569 is a better solution.

Reply:
Nvm, `libplasma` is built differently.

Reply:
@jakirkham, the `ucx` example is confusing for me as it is not clear where the cuda-specific build results are located: the `ucx-proc` packages (builds `cpu` and `gpu`) are empty and there is no build string specified for the `ucx` package whereas the packages in  https://anaconda.org/conda-forge/ucx/files seem to be cuda specific only.

For arrow-cpp, we need to support at least two scenarios:
1. using arrow-cpp from cpu-only specific software.
2. using arrow-cpp from software that can take advantage of cuda supported gpus. One can assume that the software is linked to cuda libraries.

So, there must exist at least two separate conda builds of arrow-cpp, one build with `-DARROW_CUDA=OFF` and one with `-DARROW_CUDA=ON`. Can you explain (or direct me to relevant documentation or examples) how do you foresee the above to be achievable by using mutex metapackages? As I understand, metapackages are useful for constraining solvers to specific package builds but the package builds must also exist somehow.

Another confusing matter is the choice of build strings: `gpu` is often specified to mean `cuda` but there exist `cuda` alternatives (`rocm`, for instance). While mutex metapackage works for cases with two mutually exclusive choices (`nocuda/cuda`, `mkl/nomkl`, etc), will it work for cases with more than two exclusive choices like `cuda/rocm/no-cuda-no-rocm`? Just asking out-of-curiosity...

Reply:
Is in the `ucx` version somehow a trick to ensure that if `gpu` is not requested, only the `cpu` package is installed? Triggering `conda` to prefer one version over the other is making this hard to approach for me.

Reply:
@xhochy, currently `ucx` doesn't do that, but it is possible to do that ( https://github.com/conda-forge/ucx-split-feedstock/issues/26 ).

@pearu, the `ucx` package hash changes based on whether CUDA was used and what version of CUDA was used. `ucx-proc` and `cudatoolkit` provide the constraints to force a `cpu` or `gpu` build along with CUDA version constraints.

Reply:
Thanks, @jakirkham. Is there any harm in using build strings on top of hash changes?

Reply:
The main issue with changing the `build/string` field is this replaces the whole build string (as opposed to appending it or something else less drastic). The result is it hides the build number, package hash, Python information, and anything else Conda-build would have added to the build string automatically. It's possible to include this in the manually constructed build string. However it's a bit fragile if Conda-build changes the content in build strings later. So mostly I try to avoid this in non-metapackages. Though completely up to you whether you pursue that approach. 🙂

Reply:
FWIW it is possible to inspect the package hash to see what is contained `conda inspect hash-inputs`.

cc @msarahan (who convinced me at one point to just use package hashes 😉)

Reply:
ok, gotcha.
To complete the story, how the users are able to select which package to install?
When the build string contains an only hash, will this work:
```
conda install "arrow-cpp=*=*cuda"
```
or should it be `conda install arrow-cpp arrow-cpp-build-ext=*=cuda` or smth different?

Reply:
It would be the latter. They may additionally do the following to constrain the CUDA version.

```
conda install arrow-cpp arrow-cpp-build-ext=*=cuda cudatoolkit=10.1
```

Reply:
`conda install "arrow-cpp=*=*cuda"` will work too

Reply:
It sounds like that may be removed from the build string, in which case it wouldn't.